src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && @info "Listening on: $host:$port"
 
-    if host isa Sockets.InetAddr # build debugging info
-        hostname = string(host.host)
-        hostport = string(host.port)
-    else
-        hostname = string(host)
-        hostport = ""
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(handler, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
+            if !tcpisvalid(io)
+                verbose && @info "Accept-Reject:  $io"
+                close(io)
+                continue
+            end
+            connectioncounter[] += 1
+            conn = Connection(server.hostname, server.hostport, 0, 0, true, io)
+            let io=io, count=count
+                @async begin
+                    try
+                        verbose && @info "Accept ($count):  $conn"
+                        handle(h, conn, reuse_limit, readtimeout)
+                        verbose && @info "Closed ($count):  $conn"
                     catch e
-                        @error "Error ($i):  $conn" exception=(e, stacktrace(catch_backtrace()))
+                        @error exception=(e, stacktrace(catch_backtrace()))
                     finally
                         connectioncounter[] -= 1
-                        close(conn)
-                        @info "Closed ($i):  $conn"
+                        close(io)
```

Comment:
This seems like a good change. The `Base.close(c::Connection)` was doing purging and connection pool closing that not needed for server connections. Is that your thinking? or was there another reason?

Reply:
That's right.